### PR TITLE
[MVV-2] Attach train station information to the public transport information

### PR DESCRIPTION
**Pull request by quarz12**
Fixes #

## Proposed Changes (include Screenshots if possible)

- data part of #694
- adds data from second csv, that contains train stations, too. previous pr was only bus stations

## How to test this PR

1. models/testmodels.py

## How has this been tested?

- running main.py

## Checklist

- [x] I have updated the documentation / No need to update the documentation
- [x] I have run the linter

### CODE REVIEW

File: data/external/scrapers/public_transport.py [hunk 1]
```diff
@@ -47,8 +42,59 @@ def scrape_stations():
...
+        return stations.values()
+
+
+@cached_json("public_transport.json")
+def scrape_stations():
+    """Scrape the stations from the MVV GTFS data and return them as a list of dicts"""
+    _download_file(MVV_GTFS_URL, PUBLIC_TRANSPORT_CACHE_PATH / "fahrplandaten.zip")
+    with ZipFile(PUBLIC_TRANSPORT_CACHE_PATH / "fahrplandaten.zip") as file_zip:
+        file_zip.extract("stops.txt", PUBLIC_TRANSPORT_CACHE_PATH)
+    _download_file(MVV_HST_REPORT_URL, PUBLIC_TRANSPORT_CACHE_PATH / "train_stations.csv")
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:93]: I think these files should be downloaded in the functions where they are used given that they are only used there.

**Review by CommanderStorm — CHANGES_REQUESTED** [review]

File: data/external/scrapers/public_transport.py [hunk 2]
```diff
@@ -4,16 +4,11 @@
 from external.scraping_utils import _download_file, CACHE_PATH, cached_json
 
 MVV_GTFS_URL = "https://www.mvv-muenchen.de/fileadmin/mediapool/02-Fahrplanauskunft/03-Downloads/openData/mvv_gtfs.zip"
+MVV_HST_REPORT_URL = "https://www.mvv-muenchen.de/fileadmin/mediapool/02-Fahrplanauskunft/03-Downloads/openData/MVV_HSTReport2212.csv"  # train/tram stations + some bus stations
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:7]: Above breaks the linting due to the lenght of the line
```suggestion
MVV_OPENDATA_URL = "https://www.mvv-muenchen.de/fileadmin/mediapool/02-Fahrplanauskunft/03-Downloads/openData"
MVV_GTFS_URL = f"{MVV_OPENDATA_URL}/mvv_gtfs.zip"
MVV_HST_REPORT_URL = f"{MVV_OPENDATA_URL}/MVV_HSTReport2212.csv"
```

Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:7]: This is named `MVV_HSTReport2212.csv` => Will we need to update this in the future?

File: data/external/scrapers/public_transport.py [hunk 3]
```diff
@@ -47,8 +42,59 @@ def scrape_stations():
...
+                    {
+                        "station_id": line["Globale ID"],
+                        "name": line["Name ohne Ort"],
+                        "lat": float(line["WGS84 X"].replace(",", ".")),
+                        "lon": float(line["WGS84 Y"].replace(",", ".")),
+                        "sub_stations": [],
+                    },
+                )
+            else:
+                parent_id = ".".join(line["Globale ID"].split(":")[0:3])
```
[the remark this answers is not in the document]
Comment (quarz12) [reply data/external/scrapers/public_transport.py:68]: no idea how that got there, ```:``` is correct. I even had it in earlier versions.

File: data/external/scrapers/public_transport.py (hunk 2 above)
Comment (quarz12) [reply data/external/scrapers/public_transport.py]: probably, yes

File: data/external/scrapers/public_transport.py [hunk 4]
```diff
@@ -47,8 +42,59 @@ def scrape_stations():
...
+
+@cached_json("public_transport.json")
+def scrape_stations():
+    """Scrape the stations from the MVV GTFS data and return them as a list of dicts"""
+    _download_file(MVV_GTFS_URL, PUBLIC_TRANSPORT_CACHE_PATH / "fahrplandaten.zip")
+    with ZipFile(PUBLIC_TRANSPORT_CACHE_PATH / "fahrplandaten.zip") as file_zip:
+        file_zip.extract("stops.txt", PUBLIC_TRANSPORT_CACHE_PATH)
+    _download_file(MVV_HST_REPORT_URL, PUBLIC_TRANSPORT_CACHE_PATH / "train_stations.csv")
+    stations = _load_bus_stations()
+    stations = _load_train_stations(stations)
```
[the remark this answers is not in the document]
Comment (quarz12) [reply data/external/scrapers/public_transport.py]: That was a bit messy, got to admit that. 
I have changed both methods to add to a parameter dict. It now looks like this:
```
stations = {}
_load_bus_stations(stations)
_load_train_stations(stations)
```
The reason for keeping the data in a dict is, that stations in the train data can have parents in the bus data. 
Your suggestion of merging dicts might work, too, but this one definitely does.

[the remark this answers is not in the document]
Comment (CommanderStorm) [reply data/external/scrapers/public_transport.py]: Have you checked this assumption?
I would assume this to be the other way around (bus stations belonging to train stations being a thing)

[the remark this answers is not in the document]
Comment (quarz12) [reply data/external/scrapers/public_transport.py]: in general, parent stations are neither train nor bus stations, but the collection of their substations. If you request departures for a parent, you get all the childrens departures. If a parent only has 1 child, the child becomes the parent.

Realistically, I dont think there would be a child station without its parent in either csv. That just wouldn't make sense. 
If you want, i'll change it to dict merging. But this works just as well, so, from my perspective, there's not really a point in doing so.

[the remark this answers is not in the document]
Comment (CommanderStorm) [reply data/external/scrapers/public_transport.py]: > Realistically, I dont think there would be a child station without its parent in either csv.
> That just wouldn't make sense.

Having assumptions about foreign data is what kills you in data science.
=> If you have tested that this assumption holds true, we can merge this PR. (I am fine with calling+merging it as is, debugging such an issue later will be a nightmare though => let's do so now)

[the remark this answers is not in the document]
Comment (quarz12) [reply data/external/scrapers/public_transport.py]: I added checks/warnings for this case. also found some bugs I fixed, but now all the stations are getting added to our data.

File: data/external/scrapers/public_transport.py [hunk 5]
```diff
@@ -1,54 +1,105 @@
...
+                parent["sub_stations"].append(sub_station)
+            else:
+                repeat_later.append(sub_station)
+
+    for sub in repeat_later:
+        if parent := stations.get(sub["parent"]):
+            parent["sub_stations"].append(sub)
+        else:
+            if sub["station_id"]:
+                logging.warn(f"{sub['name']} with id {sub['station_id']} has no parent in our data")
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:53]:
```suggestion
                logging.warning(f"{sub['name']} with id {sub['station_id']} has no parent in our data")
```

File: data/external/scrapers/public_transport.py [hunk 6]
```diff
@@ -1,54 +1,105 @@
...
+        else:
+            sub_station = {
+                "station_id": line["stop_id"],
+                "name": line["stop_name"],
+                "lat": float(line["stop_lat"]),
+                "lon": float(line["stop_lon"]),
+                "parent": line["parent_station"],
+            }
+            if not sub_station["parent"]:
+                sub_station["parent"]=":".join(line["stop_id"].split(":")[:3])
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:41]:
```suggestion
                sub_station["parent"] = ":".join(line["stop_id"].split(":")[:3])
```

File: data/external/scrapers/public_transport.py [hunk 7]
```diff
@@ -1,54 +1,105 @@
...
+            else:
+                repeat_later.append(sub_station)
+
+    for sub in repeat_later:
+        if parent := stations.get(sub["parent"]):
+            parent["sub_stations"].append(sub)
+        else:
+            if sub["station_id"]:
+                logging.warn(f"{sub['name']} with id {sub['station_id']} has no parent in our data")
+
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:54]:
```suggestion


```

File: data/external/scrapers/public_transport.py [hunk 8]
```diff
@@ -1,54 +1,106 @@
...
+            if parent := stations.get(parent_id):
+                parent["sub_stations"].append(sub_station)
+            else:
+                repeat_later.append(sub_station)
+    for sub in repeat_later:
+        if parent := stations.get(sub["parent"]):
+            parent["sub_stations"].append(sub)
+        else:
+            if sub["station_id"]:
+                logging.warn(f"{sub['name']} with id {sub['station_id']} has no parent in our data")
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:93]:
```suggestion
                logging.warning(f"{sub['name']} with id {sub['station_id']} has no parent in our data")
```

File: data/external/scrapers/public_transport.py [hunk 9]
```diff
@@ -1,54 +1,106 @@
 import csv
+import logging
 from zipfile import ZipFile
 
 from external.scraping_utils import _download_file, CACHE_PATH, cached_json
```
Comment (CommanderStorm) [inline data/external/scrapers/public_transport.py:5]:
```suggestion
from external.scraping_utils import _download_file, CACHE_PATH, cached_json

```

**Pull request merged**